### ISSUE TITLE
[GDNative/C++] Memory leaks in the editor

### ISSUE THREAD

**Issue by lethiandev** [godotengine/godot#19676]
**Godot version:**
3.0.3.stable.official.e649ec7
No-Mono

**OS/device including version:**
Windows 10 x64

**Issue description:**
I've been working on project with utilizing GDNative shared libraries written in C++ for significant performance gain, but I've run into some troubles with memory leaks in the editor. So I created a small project using `std::printf`s for debug purpose putting the printfs in constructors and destructors and I've found that GDNative does not call my destructors at all. In result, there is a buch of memory leaks which increases memory usage every scene save/open :/

Below is an example of the command prompt output, please notice that there is no destructors' calls printed out.
![image](https://user-images.githubusercontent.com/5323129/41665307-8ef4adb2-749f-11e8-8140-a78a8b5c98a3.png)

I'm using recent GDNative headers and bindings from [GodotNativeTools' repo](https://github.com/GodotNativeTools/godot-cpp/).

Also I've found the several other problems with GDNative, but I've decided to split them by creating different issue tickets, for convenient.

**Steps to reproduce:**
Open the editor within the provided project below and checkout the output command prompt/terminal. Try to save the `test_scene.tscn` several times and notice that there's no destructors' calls.

Sorry but I didn't provide linux version of the libraries, instead you can build them by yourself using CMake.
```bash
cd <project_dir> && mkdir build/linux && cd build/linux
cmake -G Ninja -DGDNATIVE_CPP_PATH=<cpp_bindings_dir> -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=../../lib ../..
cmake --build .
```

> If there is some problem with the build, remove entirely `build/linux` directory and re-generate build scripts again. It may help with the cached variables issue.

**Minimal reproduction project:**
[gdnative-memory-leak-bug.zip](https://github.com/godotengine/godot/files/2121017/gdnative-memory-leak-bug.zip)

**Comment by akien-mga** [comment]
CC @karroffel

**Comment by akien-mga** [comment]
Fixed by #27320.
